windows/src/developer/TIKE/xml/layoutbuilder/builder.js

```diff
@@ -2157,7 +2162,67 @@ $(function () {
     target.val(target.val() + o.text);
   };
   
+  builder.loadingState = true;
+  
+  builder.saveState = function() {
+    if(builder.loadingState) return;
+    
+    var state = {
+      platform: builder.lastPlatform,
+      layer: builder.lastLayerIndex,
+      presentation: $('#selPlatformPresentation').val()
+    };
+    
+    var key = builder.selectedKey();
+    if(key.length > 0) {
+      state.key = key.data('id');
+    }
+    
+    var subkey = builder.selectedSubKey();
+    if (subkey.length > 0) {
+      state.subkey = $(subkey).data('id');
+    }
+    
+    $.post('/app/source/toucheditor/state', {
+      'Filename': builder.filename,
+      'State': JSON.stringify(state)
+    });
+  };
+  
+  builder.loadState = function() {
+    $.get('/app/source/toucheditor/state', 
+      {
+        'Filename': builder.filename
+      },
+      function(data) {
+        builder.loadingState = false;
+        if(typeof data === 'object') {
+          if(data.platform) {
+            $('#selPlatform').val(data.platform);
+            builder.selectPlatform();
+          }
+          if(data.presentation) {
+            $('#selPlatformPresentation').val(data.presentation);
+            builder.prepareLayer();
+          }
+          if(data.layer) {
+            $('#selLayer').val(data.layer);
+            builder.selectLayer();
+          }
+          if(data.key) {
+            builder.selectKey($('#kbd .key').filter(function (index) { return $(this).data('id') === data.key; }).first());
+          }         
+          if(data.subkey) {
```

Comment:
for ll. 2212 and 2215, do you want to sanity check `$(this).data('id')` too?

Reply:
`$(this).data('id')` relates to the results in the `$('#kbd .key')` array filter, so is not relevant outside the closure.